src/db/file.rs

```diff
@@ -160,237 +31,20 @@ pub fn add_path_into_database<P: AsRef<Path>>(
-                    trans.query(
-                        "UPDATE files SET mime = $2, content = $3, date_updated = NOW() \
-                                    WHERE path = $1",
-                        &[&bucket_path, &mime, &content],
-                    )?;
-                }
-            }
-
-            file_paths_and_mimes.insert(file_path.clone(), mime.to_owned());
-        }
-
-        if !futures.is_empty() {
-            attempts += 1;
-
-            match rt.block_on(::futures::future::join_all(futures)) {
-                Ok(_) => {
-                    // this batch was successful, start another batch if there are still more files
-                    batch_size = cmp::min(to_upload.len(), MAX_CONCURRENT_UPLOADS);
-                    currently_uploading = to_upload.drain(..batch_size).collect();
-                    attempts = 0;
-                }
-                Err(err) => {
-                    error!("failed to upload to s3: {:?}", err);
-                    // if any futures error, leave `currently_uploading` in tact so that we can retry the batch
-                    if attempts > 2 {
-                        panic!("failed to upload 3 times, exiting");
-                    }
-                }
-            }
-        } else {
-            batch_size = cmp::min(to_upload.len(), MAX_CONCURRENT_UPLOADS);
-            currently_uploading = to_upload.drain(..batch_size).collect();
-        }
-    }
-
-    trans.commit()?;
-
-    let file_list_with_mimes: Vec<(String, PathBuf)> = file_paths_and_mimes
-        .into_iter()
-        .map(|(file_path, mime)| (mime, file_path))
-        .collect();
-    file_list_to_json(file_list_with_mimes)
-}
-
-fn detect_mime(file_path: &Path) -> Result<&'static str> {
-    let mime = mime_guess::from_path(file_path)
-        .first_raw()
-        .map(|m| m)
-        .unwrap_or("text/plain");
-    Ok(match mime {
-        "text/plain" | "text/troff" | "text/x-markdown" | "text/x-rust" | "text/x-toml" => {
-            match file_path.extension().and_then(OsStr::to_str) {
-                Some("md") => "text/markdown",
-                Some("rs") => "text/rust",
-                Some("markdown") => "text/markdown",
-                Some("css") => "text/css",
-                Some("toml") => "text/toml",
-                Some("js") => "application/javascript",
-                Some("json") => "application/json",
-                _ => mime,
-            }
-        }
-        "image/svg" => "image/svg+xml",
-        _ => mime,
-    })
+    let backend = Storage::new(conn);
+    let file_list = backend.store_all(conn, prefix, path.as_ref())?;
+    file_list_to_json(file_list.into_iter().collect())
 }
 
-fn file_list_to_json(file_list: Vec<(String, PathBuf)>) -> Result<Json> {
+fn file_list_to_json(file_list: Vec<(PathBuf, String)>) -> Result<Json> {
     let mut file_list_json: Vec<Json> = Vec::new();
 
     for file in file_list {
         let mut v: Vec<String> = Vec::new();
```

Comment:
Could pre-allocate this vector
```suggestion
        let mut v: Vec<String> = Vec::with_capacity(2);
```